apps/web/src/app/vault/individual-vault/vault-banners/services/vault-banners.service.ts

```diff
@@ -0,0 +1,211 @@
+import { BillingAccountProfileStateService } from "@bitwarden/common/billing/abstractions/account/billing-account-profile-state.service";
+import { PlatformUtilsService } from "@bitwarden/common/platform/abstractions/platform-utils.service";
+import { KdfType, PBKDF2_ITERATIONS } from "@bitwarden/common/platform/enums";
+import {
+  StateProvider,
+  ActiveUserState,
+  KeyDefinition,
+  PREMIUM_BANNER_DISK_LOCAL,
+  BANNERS_DISMISSED_DISK,
+} from "@bitwarden/common/platform/state";
+import { SyncService } from "@bitwarden/common/vault/abstractions/sync/sync.service.abstraction";
+
+export enum VisibleVaultBanner {
+  KDFSettings = "kdf-settings",
+  OutdatedBrowser = "outdated-browser",
+  Premium = "premium",
+  VerifyEmail = "verify-email",
+}
+
+type PremiumBannerReprompt = {
+  numberOfDismissals: number;
+  /** Timestamp representing when to show the prompt next */
+  nextPromptDate: number;
+};
+
+/** Banners that will be re-shown on a new session */
+type SessionBanners = Omit<VisibleVaultBanner, VisibleVaultBanner.Premium>;
+
+export const PREMIUM_BANNER_REPROMPT_KEY = new KeyDefinition<PremiumBannerReprompt>(
+  PREMIUM_BANNER_DISK_LOCAL,
+  "bannerReprompt",
+  {
+    deserializer: (bannerReprompt) => bannerReprompt,
+  },
+);
+
+export const BANNERS_DISMISSED_DISK_KEY = new KeyDefinition<SessionBanners[]>(
+  BANNERS_DISMISSED_DISK,
+  "bannersDismissed",
+  {
+    deserializer: (bannersDismissed) => bannersDismissed,
+  },
+);
+
+@Injectable()
+export class VaultBannersService {
+  shouldShowPremiumBanner$: Observable<boolean>;
+
+  private premiumBannerState: ActiveUserState<PremiumBannerReprompt>;
+  private sessionBannerState: ActiveUserState<SessionBanners[]>;
+
+  /**
+   * Emits when the sync service has completed a sync
+   *
+   * This is needed because `hasPremiumFromAnySource$` will emit false until the sync is completed
+   * resulting in the premium banner being shown briefly on startup when the user has access to
+   * premium features.
+   */
+  private syncCompleted$ = new Subject<void>();
+
+  constructor(
+    private tokenService: TokenService,
+    private userVerificationService: UserVerificationService,
+    private stateProvider: StateProvider,
+    private billingAccountProfileStateService: BillingAccountProfileStateService,
+    private platformUtilsService: PlatformUtilsService,
+    private kdfConfigService: KdfConfigService,
+    private syncService: SyncService,
+  ) {
+    this.pollUntilSynced();
+    this.premiumBannerState = this.stateProvider.getActive(PREMIUM_BANNER_REPROMPT_KEY);
+    this.sessionBannerState = this.stateProvider.getActive(BANNERS_DISMISSED_DISK_KEY);
+
+    this.shouldShowPremiumBanner$ = combineLatest([
+      this.billingAccountProfileStateService.hasPremiumFromAnySource$,
+      this.premiumBannerState.state$,
+      this.syncCompleted$,
+    ]).pipe(
+      map(([canAccessPremium, dismissedState]) => {
```

Comment:
```suggestion
    this.shouldShowPremiumBanner$ = combineLatest([
      this.billingAccountProfileStateService.hasPremiumFromAnySource$,
      this.premiumBannerState.state$,
    ]).pipe(
      skipUntil(() => this.syncCompleted$),
      map(([canAccessPremium, dismissedState]) => {
```

Even though `combineLatest` does work it will keep subscribed to `syncCompleted$` until it's resulting observable is unsubscribed from. With it in `skipUntil` it will unsubscribe once it got the "notification". Eventually I will add a `lastSync$` method on sync service which will be a stream of those last sync dates and you could `filter` out the `null`. I think using `skipUntil` now would make that a smaller refactor and be a little more efficient when that happens.

Reply:
@justindbaur Thanks for pointing this out, I spent probably to much time trying to work in `skipUntil` but based on the logic flow I don't think it works in this use case. 

`skipUntil` will skip the emissions from `combineLatest` until `this.syncCompleted$` emits. For this use case, `combineLatest` will usually emit before `this.syncCompleted$` and doesn't reemit again unless something were to change. I tried using `skipUntil` in combination with a couple other operators but I couldn't get the previous value from the `combineLatest` after `skipUntil` passes. Let me know if I missed something obvious here!

I ended up using a `ReplaySubject` & `take(1)` to replay the `combineLatest`. Let me know what you think! [9f43b01](https://github.com/bitwarden/clients/pull/8899/commits/9f43b0166a3961a4471d7f03caa73c1b19253486)

Reply:
@nick-livefront Thanks for giving it a try, I definitely led you astray with `skipUntil`, for whatever reason I thought it would avoid subscribing to the source observable until it emitted, then it would subscribe and the `combineLatest` would go and get it's values for the first time. None of the documentation or examples back me up on that so who knows where I got that. 

What you came up with is quite good, I think you could `mergeMap(() = premiumSource$)` directly and not need the intermediary `ReplaySubject`. But this will have the difference where those values will first get subscribed to when the sync has completed vs already going for a value and having it cached and ready to go when the sync completes. Those values should be grabbed pretty quickly though.

Reply:
For sure, to me it felt like it _should_ be a common scenario enough that it would be available with some operator. 

Anyways, one less operator the better! [fc1c4ca](https://github.com/bitwarden/clients/pull/8899/commits/fc1c4ca8f4ea28c38f40214a797d94642f69ef18)

Reply:
Yeah same, I really thought there was going to be something built in, but the `take(1), mergeMap(...)` gives the exact pattern I was hoping for. If you're at all interested [this](https://stackblitz.com/edit/typescript-spgadh?file=index.ts) is where I played around to make sure it subscribes and unsubscribes to everything in the order I was expecting. The `docClick$` is the sync completed essentially and the `timer` and `interval` are the two states that will emit at least once but not ever again necessarily. It looks perfect, thanks for doing this!